ui-speedspacechart/src/components/helpers/drawElements/speedLimitTags.ts

```diff
@@ -82,108 +85,135 @@ export const drawSpeedLimitTags = async ({
-        }
+        const iconXPosition = x + (tagWidth - ICON_BACKGROUND_WIDTH) / 2;
+        const iconYPosition = Y_POSITION - ICON_OFFSET;
+        const cornerRadius = 4;
+
+        drawRoundedRect(
+          ctx,
+          iconXPosition,
+          iconYPosition,
+          ICON_BACKGROUND_WIDTH,
+          ICON_BACKGROUND_HEIGHT,
+          cornerRadius
+        );
+
+        drawSvgImageWithColor(
+          ctx,
+          image,
+          iconXPosition + ICON_OFFSET,
+          iconYPosition + ICON_OFFSET,
+          ICON_WIDTH,
+          ICON_HEIGHT,
+          '#FFFFFF'
+        );
+      } else {
+        ctx.fillStyle = 'white';
+        ctx.font = '600 12px "IBM Plex Sans"';
+        ctx.textAlign = 'left';
+        ctx.textBaseline = 'middle';
+        const textPosition = x + TEXT_LEFT_PADDING;
+        ctx.fillText(tag, textPosition, Y_POSITION + TEXT_PADDING_TOP + RECT_HEIGHT / 2);
       }
-    });
+    }
+  });
 
-    drawSeparatorLinearLayer(ctx, 'rgba(0,0,0,0.1)', MARGINS, width, marginTop);
-    ctx.restore();
+  drawSeparatorLinearLayer(ctx, 'rgba(0,0,0,0.1)', MARGINS, width, marginTop);
+  ctx.restore();
 
-    // prevent overlapping with margins left and right
-    ctx.clearRect(0, 0, MARGIN_LEFT, LINEAR_LAYERS_HEIGHTS.SPEED_LIMIT_TAGS_HEIGHT);
-    ctx.clearRect(width - MARGIN_RIGHT, 0, width, LINEAR_LAYERS_HEIGHTS.SPEED_LIMIT_TAGS_HEIGHT);
+  // prevent overlapping with margins left and right
+  ctx.clearRect(0, 0, MARGIN_LEFT, LINEAR_LAYERS_HEIGHTS.SPEED_LIMIT_TAGS_HEIGHT);
+  ctx.clearRect(width - MARGIN_RIGHT, 0, width, LINEAR_LAYERS_HEIGHTS.SPEED_LIMIT_TAGS_HEIGHT);
+};
+
+export const computeTooltip = async ({
+  width,
+  height: marginTop,
+  store,
+}: DrawFunctionParams): Promise<tooltipInfos | null> => {
+  const { speedLimitTags, ratioX, leftOffset, cursor } = store;
+
+  const { MARGIN_TOP, MARGIN_LEFT } = MARGINS;
+
+  const maxPosition = maxPositionValue(store.speeds);
+
+  let tooltip: tooltipInfos | null = null;
 
-    return tooltip;
+  if (speedLimitTags == undefined) return null;
+
+  for (const { position, value } of speedLimitTags) {
+    const tooltipTextMap: Record<string, string> = {
+      incompatible: 'Incompatible with the infrastructure',
+    };
+
+    const { tag } = value;
+    if (!(tag in tooltipTextMap)) continue;
+    const potential_tooltip_text = tooltipTextMap[tag];
+
+    const x = positionOnGraphScale(position.start, maxPosition, width, ratioX, MARGINS);
+    const nextBoundary = positionOnGraphScale(position.end!, maxPosition, width, ratioX, MARGINS);
+
+    if (nextBoundary == null || cursor.x == null || cursor.y == null) break;
+
+    const tagWidth = nextBoundary - x - RECTANGLE_SPACING;
+    if (
+      cursor.x >= x - MARGIN_LEFT + leftOffset &&
+      cursor.x <= x - MARGIN_LEFT + leftOffset + tagWidth &&
+      cursor.y >= marginTop - MARGIN_TOP + Y_POSITION - 2 &&
+      cursor.y <= marginTop - MARGIN_TOP + Y_POSITION - 1 + RECT_HEIGHT
+    ) {
+      tooltip = {
```

Comment:
I think we can straight up `return` here without the intermediate variable?

Reply:
Indeed we can, thanks